File: 프로그래머스/3/150367. 표현 가능한 이진트리/표현 가능한 이진트리.py
def toBinary(n):
    result = ""
    while n > 1:
        result += str(n % 2)
        n //= 2
    
    result += str(n)
    
    return result[::-1]

def divide(n):
    stack = [n]
    
    while stack:
        cur = stack.pop()
            
        half = len(cur) // 2
        
        if cur[half] == "0":
            return 0
        
        right = cur[:half]
        left = cur[half+1:]
        
        if len(cur) > 3:
            if len(set(right)) > 1:
                stack.append(right)
            
            if len(set(left)) > 1:
                stack.append(left)
                
    return 1

def solution(numbers):
    answer = []
    
    for i in numbers:
        binary = toBinary(i)
        length = len(binary)
        n = 1
        
        while True:
            if 2**(n) - 1 < length:
                n += 1
            else:
                break
        
        binary = "0" * ((2**n - 1) - length) + binary
        
        answer.append(divide(binary))
    
    return answer

File: 프로그래머스/3/150367. 표현 가능한 이진트리/표현 가능한 이진트리.java
import java.util.ArrayDeque;

class Solution {
    
    public int divide(String n) {
        ArrayDeque<String> stack = new ArrayDeque<>();
        stack.push(n);
        
        while (stack.size() > 0) {
            String cur = stack.pop();
            int half = cur.length() / 2;
            
            if (cur.charAt(half) == '0') return 0;
            
            String left = cur.substring(0, half);
            String right = cur.substring(half+1);
            
            if (cur.length() > 3) {
                if (left.contains("1")) stack.push(left);
                
                if (right.contains("1")) stack.push(right);
            }      
        }
        return 1;
    }
    
    public int[] solution(long[] numbers) {
        int length = numbers.length;
        int[] answer = new int[length];
        
        for (int i = 0; i < length; i++) {
            String binary = Long.toString(numbers[i], 2);
            int start = 1;
            int size = binary.length();
            
            while (true) {
                int curPow = (int) Math.pow(2, start) - 1;
                
                if (curPow >= size) {
                    binary = "0".repeat(curPow - size) + binary;
                    break;
                }
                else{
                    start += 1;
                }
            }              
            answer[i] = divide(binary);
        }        
        return answer;
    }
}